File: Algorithm/Programmers/level2_hash_위장.py
"""
의상들을 입을 수 있고 안입을 수 있어서 +1,
의상을 안걸치는 경우는 없기에 -1

(의상1의 수+1) * (의상2의 수+1) ... -1

"""


def solution(clothes):
    answer = 1
    key = {x[1] : 0 for x in clothes}
    
    for c in clothes :
        key[c[1]] += 1
    
    for n in key.values() :
        answer *= (n+1)
    
    return answer-1

File: Algorithm/Programmers/java/level2_hash_위장.java
/*
프로그래머스 고득점kit - hash level2

https://school.programmers.co.kr/learn/courses/30/lessons/42578

clothes	return
[["yellow_hat", "headgear"], ["blue_sunglasses", "eyewear"], ["green_turban", "headgear"]]	5
[["crow_mask", "face"], ["blue_sunglasses", "face"], ["smoky_makeup", "face"]]	3

*/

import java.util.*;

class Solution {
    public int solution(String[][] clothes) {
        int answer = 1;
        Map<String,Integer> cMap = new HashMap();
        
        for(String[] c : clothes){
            cMap.put(c[1],cMap.getOrDefault(c[1],0) + 1);
        }
                
        for(String key : cMap.keySet()){
            answer *= cMap.get(key) + 1;
        }
        
        return answer-1;
    }
}


//update 2024.07.14
import java.util.*;

class Solution {
    public int solution(String[][] clothes) {
        int answer = 0;
        Map<String, Integer> maps = new HashMap<>();
        
        for(String[] first : clothes){
            maps.put(first[1], maps.getOrDefault(first[1], 1) + 1);
        }
        
        return maps.values().stream().reduce(1, (a,b) -> a*b) - 1;
    }
}


